### PR TITLE
Build Cmd Unit Tests

### PR DESCRIPTION
**Pull request by astarinmymind**

### CODE REVIEW

File: src/cmds/eval/evalWorker.ts [hunk 1]
```diff
@@ -22,6 +22,13 @@ if (parentPort !== null) {
     compartment.evaluate(
       // '() => ' + readFileSync(pluginFilePath, 'utf8')
       readFileSync(pluginFilePath, 'utf8'),
+      {
+        sloppyGlobalsMode: true,
+        __moduleShimLexicals__: undefined,
+        __evadeHtmlCommentTest__: false,
+        __evadeImportExpressionTest__: true,
+        __rejectSomeDirectEvalExpressions__: false,
+      },
```
Comment (rekmarks) [inline src/cmds/eval/evalWorker.ts:31]: We aren't going to use these options in prod, so this entire config is going to have to go, and (some of) the transforms re-enabled.

However, we should PR the changes to the transforms separately, so for now, we can do whatever works. I'd prefer if we deleted these lines and uncommented the transforms in `postProcess`, if possible.

File: test/utils/misc.test.js [hunk 2]
```diff
@@ -201,6 +212,37 @@ describe('misc', () => {
...
+      });
+      jest.spyOn(fs.promises, 'unlink').mockRejectedValueOnce();
+      const errorMock = jest.spyOn(console, 'error').mockImplementation();
+      await expect(writeError('foo', 'bar', new Error('error message'), 'dest'))
+        .rejects
+        .toThrow('process exited');
+      expect(errorMock).toHaveBeenNthCalledWith(1, 'foo bar');
+      expect(errorMock).toHaveBeenCalledTimes(2, `foo Failed to unlink mangled file.`);
+    });
+  });
```
Comment (rekmarks) [inline test/utils/misc.test.js:244]: This could use some breathing space.

File: src/utils/misc.ts [hunk 3]
```diff
@@ -79,6 +80,35 @@ export function logWarning(msg: string, error?: Error): void {
...
+  } catch (unlinkError) {
+    logError(`${processedPrefix}Failed to unlink mangled file.`, unlinkError);
+  }
+
+  // unless the watcher is active, exit
+  if (!global.snaps.isWatching) {
+    process.exit(1);
+  }
+}
+
```
Comment (rekmarks) [inline src/utils/misc.ts:111]: What motivated this move?

**Review by rekmarks — CHANGES_REQUESTED** [review]
We can see the light at the end of the tunnel!

In addition to my inline comments, `canCloseStream` is currently not tested at all.

File: src/utils/misc.ts (hunk 3 above)
Comment (astarinmymind) [reply src/utils/misc.ts:111]: I felt it was fitting to be in the utility section, and I put it here so I could test it. Unit testing is hard when functions in the same module use each other, as I can't mock/spyOn them.

File: src/cmds/build/bundleUtils.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  let processedString = bundleString.trim();
+
+  if (options.stripComments) {
+    processedString = stripComments(processedString);
+  }
+
+  // // .import( => ["import"](
+  // processedString = processedString.replace(/\.import\(/gu, '["import"](');
+
```
Comment (rekmarks) [inline src/cmds/build/bundleUtils.ts:62]: This can be deleted.

Comment (rekmarks) [inline src/cmds/build/bundleUtils.ts:62]: This can be deleted.

File: src/cmds/build/bundleUtils.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  // if we don't do the above, the below causes syntax errors if it encounters
+  // things of the form: "something.eval(stuff)"
+  // eval(stuff) => (1, eval)(stuff)
+  processedString = processedString.replace(/(\b)(eval)(\([^)]*\))/gu, '$1(1, $2)$3');
+
+  // // SES interprets syntactically valid JavaScript '<!--' and '-->' as illegal
+  // // HTML comment syntax.
+  // // '<!--' => '<! --' && '-->' => '-- >'
+  // processedString = processedString.replace(/<!--/gu, '< !--');
+  // processedString = processedString.replace(/-->/gu, '-- >');
```
Comment (rekmarks) [inline src/cmds/build/bundleUtils.ts:78]: This can be deleted.

File: src/cmds/build/bundleUtils.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  //   throw new Error(
+  //     `Bundled code is empty after postprocessing.`,
+  //   );
+  // }
+
+  // // handle some cases by declaring missing globals
+  // // Babel regeneratorRuntime
+  // if (processedString.indexOf('regeneratorRuntime') !== -1) {
+  //   processedString = `var regeneratorRuntime;\n${processedString}`;
+  // }
```
Comment (rekmarks) [inline src/cmds/build/bundleUtils.ts:95]: These should not be commented.

File: src/cmds/build/bundleUtils.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+export function postProcess(bundleString: string | null, options: Option): string | null {
+
+  if (typeof bundleString !== 'string') {
+    return null;
+  }
+
+  let processedString = bundleString.trim();
+
+  if (options.stripComments) {
+    processedString = stripComments(processedString);
```
Comment (rekmarks) [inline src/cmds/build/bundleUtils.ts:57]: This line isn't tested, but should be.

**Review by rekmarks — CHANGES_REQUESTED** [review]
These tests look good, just some requests regarding untested / incorrectly commented code.

File: src/cmds/build/bundleUtils.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+  if (processedString.length === 0) {
+    throw new Error(`Bundled code is empty after postprocessing.`);
+  }
+
+  // handle some cases by declaring missing globals
+  // Babel regeneratorRuntime
+  if (processedString.indexOf('regeneratorRuntime') !== -1) {
+    processedString = `var regeneratorRuntime;\n${processedString}`;
+  }
```
Comment (rekmarks) [inline src/cmds/build/bundleUtils.ts:96]: We're missing test cases for these.

For the first one, we just need to confirm that `postProcess` throws an error if the processed string is of length 0. I think passing in a string consisting of nothing but spaces would do it.

For the second one, we should check that the resulting string becomes prefixed with `var regeneratorRuntime;\n`.

**Review by rekmarks — CHANGES_REQUESTED** [review]
See inline request.

Note that I ran `prettier` on all files in the PR to fix a number of style / formatting issues.

File: src/cmds/build/bundleUtils.ts (hunk 8 above)
Comment (rekmarks) [reply src/cmds/build/bundleUtils.ts:96]: I fixed this here: 4b439fe

**Review by rekmarks — APPROVED** [review]

**Pull request merged**